subsys/net/lib/nrf_cloud/src/nrf_cloud_agps.c

```diff
@@ -583,6 +583,8 @@ int nrf_cloud_agps_process(const char *buf, size_t buf_len)
 		if (!err) {
 			LOG_ERR("nRF Cloud returned A-GPS error: %d", nrf_err);
 			return -EFAULT;
+		} else { /* JSON message received but no valid error code found */
```

Comment:
`nrf_cloud_handle_error_message` returns `-ENOMSG` if the data provided was not JSON, so this comment is not quite correct.  Also, in that case, do not break... since the data was not JSON it should be check for AGPS data.

Reply:
There's a check above that the data starts with `'{'`, so the implementation assumes at this point that data is JSON and not A-GPS data. Because `nrf_cloud_handle_error_message()` first checks if the data is valid JSON, maybe the `'{'` check could be removed?

- If `nrf_cloud_handle_error_message()` returns `0` the data was an error message for A-GPS, return `-EFAULT`
- If `nrf_cloud_handle_error_message()` returns `-ENOMSG` the data was not JSON, try to parse it as A-GPS data
- Else data was JSON, don't try to parse it and return

Reply:
I checked what exactly happens when I establish the cloud connection. There is a NRF_CLOUD_EVT_RX_DATA event with payload like this: 
{"desired":{"pairing":{"state":"paired","topics":{"d2c":"prod/ec9dc0a2-736f-4899-8fd3-c9446b1c2822/m/d/nrf-351358811331336/d2c","c2d":"prod/ec9dc0a2-736f-4899-8fd3-c9446b1c2822/m/d/nrf-351358811331336/c2d"}},"nrfcloud_mqtt_topic_prefix":"prod/ec9dc0a2-736f-4899-8fd3-c9446b1c2822/"},"config":{"activeMode":true,"gnssTimeout":60,"activeWaitTime":120,"movementResolution":120,"movementTimeout":3600,"movementThreshold":10,"nod":[]}}

So it is quite valid JSON. But obviously it does not contain NRF_CLOUD_JSON_ERR_KEY because it has nothing to do with A-GPS and get_error_code_value() returns -ENOMSG. Therefore also nrf_cloud_handle_error_message() returns -ENOMSG. And if I don't return from nrf_cloud_agps_process() at this point, the A-GPS data parsing will start and error message is shown.

I would change get_error_code_value() so that it also returns -EBADMSG in case no NRF_CLOUD_JSON_ERR_KEY is found.

I agree with Tommi that '{' could be removed.

Reply:
ah, yes i see.
can you leave `get_error_code_value` unchanged and instead change the return value in `nrf_cloud_handle_error_message` as follows?

```
if (!root_obj) {
		LOG_DBG("No JSON found");
		return -ENODATA;
}
```

and add the new value to the function comment:
`* @retval -ENODATA JSON data was not found.`

also, the check for '{' is just a crude way to avoid unnecessary calls to `cJSON_Parse`.

Reply:
Be careful modifying this API.  Be sure to review all callers for both MQTT and REST -- that check for '{' could still be necessary for certain applications or samples with certain Kconfig settings, like for P-GPS + A-GPS or other combinations.

Reply:
Good comment, I'll add @jarlamsa as he appears to have worked on location_assist_write_cb() recently.

Correct, we don't have any samples using location_agps_data_process() at the moment. But I just realized @glarsennordic could have called it in his mqtt_multi_service_sample instead of calling nrf_cloud_agps_process() directly as location lib has this interface exactly for this purpose.